src/librustc_mir/interpret/eval_context.rs

```diff
@@ -546,57 +566,143 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         }
     }
 
-    pub(super) fn pop_stack_frame(&mut self) -> EvalResult<'tcx> {
+
+    pub(super) fn pop_stack_frame_internal(&mut self) -> EvalResult<'tcx, StackPopCleanup> {
+        use UnwindingState::*;
         info!("LEAVING({}) {}", self.cur_frame(), self.frame().instance);
         ::log_settings::settings().indentation -= 1;
         let frame = self.stack.pop().expect(
             "tried to pop a stack frame, but there were none",
         );
-        M::stack_pop(self, frame.extra)?;
+        let stack_pop_info = M::stack_pop(self, frame.extra)?;
+        let prev = self.unwinding;
+
+        // Handle all possible state transitions
+        self.unwinding = match (self.unwinding, stack_pop_info.unwinding) {
+            // If we weren't previously unwinding, we start performing
+            // unwinding (not in cleanup, since unwinding never starts at a
+            // cleanup block)
+            (NotUnwinding, true) => Unwinding { in_cleanup: false },
+            // Nothing special here - we weren't unwinding before, and nothing changed
+            (NotUnwinding, false) => NotUnwinding,
+            // We were previously unwinding, and we want to keep unwinding
+            (Unwinding { in_cleanup: true }, true) => Unwinding { in_cleanup: true },
+            // We should never stop unwinding while in a cleanup block - cleanup
+            // blocks are generated by the compiler, and should never 'catch'
+            // an unwind
+            (Unwinding { in_cleanup: true }, false) => {
+                panic!("Attempted to stop unwinding during cleanup!")
+            },
+            // We want to keep unwinding - our cleanup state doesn't change,
+            // since we haven't hit a terminator with a cleanup block
+            (Unwinding { in_cleanup: false}, true) => Unwinding { in_cleanup: false },
+            // The machine has indicated that unwinding is now complete
+            (Unwinding { in_cleanup: false}, false) => NotUnwinding
+        };
+        info!("unwinding: state before {:?} from stack_pop {:?} after {:?}",
+              prev, stack_pop_info.unwinding, self.unwinding);
+
         // Abort early if we do not want to clean up: We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
         match frame.return_to_block {
-            StackPopCleanup::Goto(_) => {},
-            StackPopCleanup::None { cleanup } => {
+            StackPopCleanup::Goto{ .. } => {},
+            StackPopCleanup::None { cleanup, .. } => {
                 if !cleanup {
                     assert!(self.stack.is_empty(), "only the topmost frame should ever be leaked");
                     // Leak the locals, skip validation.
-                    return Ok(());
+                    return Ok(frame.return_to_block);
```

Comment:
This should `bug!` if we are unwinding (that goes with what I said about about not ever unwinding across a `StackPopCleanup::None`).

Reply:
I've added an assertion above